### PR TITLE
Enable MC absorption for fixed window scans

### LINKED ISSUE #27459: CalculateMonteCarloAbsorption for elastic scans
MonteCarlo absorption normally requires an input workspace in the x-units of wavelength.
However, for the elastic (and inelastic) fixed energy window scans, the x-unit is the scanning observable (e.g. temperature).
Currently the `CalculateMonteCarloAbsorption` transposes the workspace, if the x-axis is not in wavelength, assuming that the y-axis is in wavelength. It also supports the `_elastic` instrument definitions for (OS) IRIS.
To be able to handle also the ILL reduced data, the logic should be generalised as follows:
- If x unit is wavelength, go straight to calculation
- Else if it is in energy transfer or energy, convert to wavelength
- Else, take the Efixed from the instrument parameters, generate a dummy workspace in the unit of wavelength with a single bin corresponding to the wavelength of Efixed, assuming elastic scattering.

### PR DESCRIPTION
**Pull request by gvardany**
**Description of work.**
This PR makes it possible to calculate MC based absorption corrections also for elastic and inelastic fixed window scans with IN16B at the ILL. In both cases, it will calculate the correction only for the lambda fixed assuming elastic scattering.
Also a bug has been fixed in the case of annular sample in a container.

**To test:**
Use the newly added test file:
```
fws = Load('ILL_IN16B_FWS_Reduced.nxs')
CalculateMonteCarloAbsorption(SampleWorkspace=mtd['fws'][0],
                              Height=1,
                              SampleWidth=1,
                              SampleThickness=0.1,
                              SampleAngle=115,
                              SampleDensity=1.5,
                              SampleChemicalFormula='H2-O',
                              CorrectionsWorkspace='corr_2theta')
CalculateMonteCarloAbsorption(SampleWorkspace=mtd['fws'][1],
                              Height=1,
                              SampleWidth=1,
                              SampleThickness=0.1,
                              SampleAngle=115,
                              SampleDensity=1.5,
                              SampleChemicalFormula='H2-O',
                              CorrectionsWorkspace='corr_q')
CalculateMonteCarloAbsorption(SampleWorkspace=mtd['fws'][2],
                              Height=1,
                              SampleWidth=1,
                              SampleThickness=0.1,
                              SampleAngle=115,
                              SampleDensity=1.5,
                              SampleChemicalFormula='H2-O',
                              CorrectionsWorkspace='corr_red')
```
All the output correction workspaces should have x-axis unit as wavelength with the value of 6.2626 AA.
The spectrum (vertical) axis should be in 2theta for *corr_2theta*, q for *corr_q*, and spectrum number for *corr_red*.

Fixes #27459, #28347 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've not had a look in here before. I originally thought this was a simple wrapper around the MonteCarloAbsorption algorithm but I can see there's a bit more going on!

I ran the test and works fine.

I've added one detail level request against the unit test. In addition a couple of comments\questions:
1) the method of identifying the ISIS instruments (two instrument names specified) doesn't seem v robust but I note that this pre-dates this change. Please could you add the two ISIS instrument names (IRIS and OSIRIS) used into the documentation?
2) Please could you add some comments to the documentation to explain what shape the container is for each of the sample shapes eg if the user supplies Shape=Annulus then the sample is an annulus and the container is two parts - an annulus inside the sample and another annulus outside the sample (or something similar). Maybe I was a bit slow but took me a while to understand what was going on with the shapes

**Comment by gvardany** [comment]
This algorithm will be deprecated in PR #28462 , so I did not bother to clarify the documentation here, but will do there for the new algorithm directly. For the 2 ISIS instruments, I am actually not happy that this code is in the algorithm; it should rather be somewhere out in ISIS specific workflows, but for the moment I kept it. Do you have a better idea where that logic could be better placed?

**Comment by DannyHindson** [comment]
I don't know the answer to your question about where the logic about the ISIS instruments should live. I will ask @AnthonyLim23 about this but I think he's on holiday this week.

A few other thoughts:
a) there are currently only two instruments that have an instrument definition file that is called *__elastic_Definition.xml. So the code could check for the presence of an IDF with a filename ending with this string and apply the ISIS specific logic that way. Instead of hardcoding the ISIS\OSIRIS instrument names
b) we could add an extra setting to the instrument definition file
c) we could add an extra on\off parameter to trigger this behaviour in the CalculateMonteCarloAbsorption algorithm but this would then need any upstream scripts to be updated. Assuming there are more than 2 non-ISIS instruments who want to use this algorithm you could set the default value for ISIS behaviour to be "off" to try and make it more backwards compatible. I see there is a UI (see Interfaces, Indirect, Corrections) that calls CalculateMonteCarloAbsorption so this would probably also need an extra checkbox adding to it

In summary I think this is a question we will need to follow up on outside of this PR

**Review by DannyHindson — APPROVED** [review]
Thanks for changing the test. Comment noted about the documentation. All looks good now

**Comment by gvardany** [comment]
I agree, I never understood the need for `*__elastic_Definition.xml`s in the first place; here I just tried not to break it. We shall hear from Anthony outside this PR.
